Build section for portable consent feature - alpha release (#589)
xmtp/xmtp-dot-org: docs/tutorials/portable-consent.md

For example, if your user blocked a contact on a different device or in another app and then uses your app, you can load the latest consent list to be sure to not mistakenly display conversations with the blocked contact.

## Understand contact consent states

A contact can have one of three consent states:

- Unknown
- Allowed
- Blocked

Here are some of the ways contact consent states are set:

**Unknown**

- For a new conversation that a new contact wants to start with your user, the contact consent state is set to `unknown`.
    
    For example, you can consider displaying a conversation with an `unknown` contact on a **Requests** tab and give your user the option to block or allow the contact.
    
- For a preexisting conversation with only messages from a contact that hasn’t had its consent state updated on the network, the contact consent state is set to `unknown` by default.
    
    For example, you can consider displaying a conversation with an `unknown` contact on a **Requests** tab and give your user the option to block or allow the contact.
    
**Allowed**

> - For a new conversation that your user requests with a contact, the contact consent state is set to `allowed`. Your user’s creation of the conversation with the contact is considered as consenting to the conversation with the contact.

Reviewer: A user might click a subscribe button or turn on notifications in a dapp, which would set their consent state to allowed for that sender.

>     For example, you can consider displaying the conversation with the `allowed` contact on the **Messages** tab.

Reviewer: I think offering this strikes the right tone -- but I'm wondering if there's something we should add to stress to the app developer that Allow/Block is an expression of user intent that needs to be respected?

So these examples are framed as suggestions for how to respect user intent.

Author: I totally get it now, Saul - thank you for the gentle guidance. I'm working on an update that will frame these examples as suggestions for how to respect user intent.

> - For a preexisting conversation with one or more messages from a contact that hasn’t had its consent state updated on the network (`unknown`) and your user has taken the action to send a message in response, update the contact consent state to `allowed`. You can consider the response from your user as consenting to the conversation with the contact.

Author: I wonder if this needs to more explicitly state that the contact consent state should be updated in the consent list on the network?

Same for the Allow and Block actions below.

Reviewer 2: Well I think there is two scenarios here. One is a preexisting conversations that pre-dates this consent work. I think thats the case being handled here. The other scenario is a new conversation on a device that doesn't have any consent type functionality yet. In that case I think the network should handle it.

Author: I'm trying something with the update to account for this... Let's see if it works! ;-) More soon...

    
    For example, you can consider moving the conversation with the contact from the **Requests** tab to the **Messages** tab.
    
- For a contact that your user has taken the action to allow, update the contact consent state to `allowed`.

    For example, you can consider displaying the conversation with the allowed contact on the **Messages** tab in your user’s inbox.

**Blocked**

- For a contact that your user has taken the action to block, update the contact consent state to `blocked`.
    
    For example, you can consider removing a conversation with the blocked contact from your user’s inbox completely.
    
## Build portable consent

Use the following methods to build portable consent in your app.

### Block or allow contacts

> To enable your user to block or allow a contact, call the following methods. Note that these functions accept lists, so you can do batch requests.

Author: Do we want to add code samples beyond TS?

Reviewer 2: FWIW these code samples are React Native. I can provide samples for iOS and Andriod as well. @rygine could probably provide for the React, JS side.

Author: I've added placeholders for Fabri to provide the JS code samples. We can always publish and just keep adding code samples for more languages as they become available. Thank you so much!

```

### Refresh the contact consent list

To ensure that you’re using the latest contact consent state, make sure to refresh the consent list from the network. Perform the refresh just in case the consent state has changed on a different device, for example.

```tsx
client.contacts.refreshConsentList();
```

### Check if a contact is blocked or allowed

Call the following methods to check if a contact is blocked or allowed.

```tsx
client.contacts.isAllowed(wantedConvo.peerAddress);
client.contacts.isBlocked(spamConvo.peerAddress);
```

### Get a conversation’s consent state

When loading a list of conversations, take into account its current consent state. You can get the `consentState` directly from the conversation.

For example, you might want to display only allowed conversations (conversations with allowed contacts) on a **Messages** tab and “unknown” conversations (conversations with unknown contacts) on a **Requests** tab. For blocked conversations (conversations with blocked contacts), you might want to not display them at all.

```tsx
const state = await conversation.consentState();
if (state === "blocked) {
	// hide the conversation
}
```

> ## Synchronize portable consent

Reviewer: Does this section apply for all devs, or only devs who already have local consent data?

Reviewer 2: All devs who want to use consent should adhere by these rules.

Reviewer: Got it. The flow chart provides necessary clarity but is also pretty intimidating in terms of the amount of integration complexity it implies.

This diagram illustrates the logic for how contact consent states are set in an app and the consent list on the XMTP network. 

> Your app must not overwrite a contact consent state in the consent list on the network except if your user performs one of the following explicit actions:

Author: Just making sure I've worded these correctly and that they cover the select list of actions that should update the consent list.

- Blocks a contact. Update the contact consent state to `blocked`.
- Allows a contact. Update the contact consent state to `allowed`.

> - Sends their first response message to a preexisting conversation requested by an unknown contact. Update the contact consent state from `unknown` to `allowed`.

Reviewer 2: > Sends their first response message to a preexisting conversation requested by an unknown contact. Update the contact consent state from `unknown` to `allowed`.

I think this should happen on the network by default. 
The scenarios I see are:

- The user explicitly clicks a block button
- The user explicitly clicks a allow button
- A preexisting conversation with a `unknown` consent state but a legacy consent state in the local DB exists. The consent state should be updated to match the legacy local state.
- A preexisting conversation that has a pre existing message from the user already --- can be updated `allowed`

A contact can have one of three consent states:

- Unknown
- Allowed
- Blocked

Here are some of the ways contact consent states are set:

**Unknown**

- For a new conversation that a new contact wants to start with your user, the contact consent state is set to `unknown`.
    
    For example, you can consider displaying a conversation with an `unknown` contact on a **Requests** tab and give your user the option to block or allow the contact.
    
- For a preexisting conversation with only messages from a contact that hasn’t had its consent state updated on the network, the contact consent state is set to `unknown` by default.
    
    For example, you can consider displaying a conversation with an `unknown` contact on a **Requests** tab and give your user the option to block or allow the contact.
    

**Allowed**

- For a new conversation that your user requests with a contact, the contact consent state is set to `allowed`. Your user’s creation of the conversation with the contact is considered as consenting to the conversation with the contact.
    
    For example, you can consider displaying the conversation with the `allowed` contact on the **Messages** tab.
    

> - For a preexisting conversation with one or more messages from a contact that hasn’t had its consent state updated on the network (`unknown`) and at least one reply from your user, update the contact consent state to `allowed`. You can consider the reply from your user as consenting to the conversation with the contact.

Author: Regarding:

>update the contact consent state to `allowed`

Just making sure - this means that this is a case in which the app SHOULD update the consent list on the network?

    
    For example, you can consider moving the conversation with the contact from the **Requests** tab to the **Messages** tab.
    
- For a contact that your user has taken the action to allow, update the contact consent state to `allowed`.

    For example, you can consider displaying the conversation with the allowed contact on the **Messages** tab in your user’s inbox.

**Blocked**

- For a contact that your user has taken the action to block, update the contact consent state to `blocked`.
    
    For example, you can consider removing a conversation with the blocked contact from your user’s inbox completely.
    
## Build portable consent

Use the following methods to build portable consent in your app.

### Block or allow contacts

To enable your user to block or allow a contact, call the following methods. Note that these functions accept lists, so you can do batch requests.

<Tabs groupId="sdk-langs">
<TabItem value="js" label="JavaScript"  attributes={{className: "js_tab"}}>

```js

> <Placeholder for JS from Fabri>

Author: ```suggestion
paste code sample here
```

---
sidebar_label: Portable consent
sidebar_position: 5.5
hide_table_of_contents: false
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import requeststab from '/docs/tutorials/img/requests-tab.png';
import messagestab from '/docs/tutorials/img/messages-tab.png';
import consentlogic from '/docs/tutorials/img/consent-state-logic.png';

# Use portable consent for XMTP

![Feature status](https://img.shields.io/badge/Feature_status-Alpha-orange)

:::caution This feature is in **alpha** status

> This feature is in **alpha** status and ready for you to start experimenting with. However, we do not recommend using alpha features in production apps. Expect frequent changes as we iterate based on feedback.

Author: GREAT idea!

Linking here: https://github.com/orgs/xmtp/discussions/49

:::

When you build an app with XMTP, you can choose to implement portable consent. 

With portable consent, a blockchain account address registered on the XMTP network can have one of three contact consent states:

- Unknown
- Allowed
- Denied

Contact consent states are stored in a consent list on the XMTP network. The consent list is accessible by all apps.

## Understand contact consent states

Here are some of the ways contact consent states are set:

### Unknown

Conversation created using an app **with** portable consent implemented:

- For a new conversation that a peer contact wants to start with a user, the contact consent state is set to `unknown`.

Conversation created using an app **without** portable consent implemented:

> - For an existing conversation created by a peer contact and that a user hasn't responded to, the contact consent state is set to `unknown`.

Reviewer 2: This is a bit confusing to me but
- a app is on a version of the sdk that supports consent (nothing on implemented since there is no action needed from the integrator other than being on a sdk version that supports it) the state is `unknown` for new conversations created by a peer
- or a app is on a old version of the sdk that doesn't have support for consent. then all conversations will have the state of unknown irregardless of who creates them.

Author: Whew - thank you so much for this clarification - I think I get it. Working on an update now!

### Allowed

> Conversation created using an app **with** portable consent implemented:

Reviewer 2: Not necessarily about implemented but the app is on a sdk version that supports consent. Since this is implicit without needed action by the developer.

- For a new conversation that a user created with a peer contact, the SDK sets the contact consent state to `allowed`.  

  The user’s creation of the conversation with the contact is considered consent.

> - For an existing conversation created by a peer contact that hasn’t had its consent state updated on the network (`unknown`) and that the user has responded to, the app must update the contact consent state to `allowed`.  
> 

Reviewer 2: This update was approved in the discussion

```suggestion
- For an existing conversation created by a peer contact that hasn’t had its consent state updated on the network (`unknown`) and that the user has responds to, the network will update contact consent state to `allowed`.  

```

  The user's response to the conversation is considered consent.

- For a peer contact that a user has taken the action to allow, subscribe to, or enable notifications from, for example, the app must update the contact consent state to `allowed`.

> Conversation created using an app **without** portable consent implemented:

Reviewer 2: If the app has not updated to the sdk version that supports consent the consent state will actually be `unknown`. This is assuming the app is on the sdk version that supports consent FWIW

- For a new conversation that a user created with a peer contact, the SDK sets the contact consent state to `allowed`.  

  The user’s creation of the conversation with the contact is considered consent.

- For an existing conversation created by a peer contact that the user has responded to, the SDK sets the contact consent state is set to `allowed`.  

  The user's response to the conversation is considered consent.

### Denied

Conversation created using an app **with** portable consent implemented:

- For a peer contact that a user has taken the action to block, unsubscribe from, or disable notifications from, for example, the app must update the contact consent state to `denied`.

## Handle consent states to respect user intent

Your app should aim to handle contact consent states appropriately because they are an expression of user intent.

> For example, if a user blocked a contact, your app should respect the user's intent to not see messages from the blocked contact. Handling the contact consent state incorrectly and showing the user messages from the blocked contact may cause the user to lose trust in your app.

Reviewer 2: ❤️
